src/language-js/utils.js

```diff
@@ -11,6 +11,21 @@ const {
 } = require("../common/util");
 const handleComments = require("./comments");
 
+/**
+ * @typedef {import("estree").BaseNode} BaseNode
+ * @typedef {import("estree").BaseExpression} BaseExpression
+ * @typedef {import("estree").Expression} Expression
+ *
+ * @typedef {Object} MemberExpressionNode
```

Comment:
There should be `MemberExpression` in `estree`

Reply:
I recall having some trouble using `MemberExpression`, will give it another try.

Reply:
Maybe because `optional`? Or `typeAnnotation`? In that case, maybe we should use babel or eslint-typescript AST Type

Reply:
Using `MemberExpression` gave me type errors like this:

```
src/language-js/utils.js:912:6 - error TS2367: This condition will always return 'false' since the types '"MemberExpression"' and '"Literal"' have no overlap.

912     (node.type === "Literal" && node.regex && node.regex.pattern) ||
```

I did change it to mix in `MemberExpression` and was able to get rid of the extra Expression properties.

I can give `@babel/types` a try. The eslint-typescript AST type seems to be specific to Typescript, or am I missing something there?

Reply:
Both `eslint-typescript` and  `babel` has own AST  type , and  some of them  is  nit  standard estree node.

`eslint-typescript` mean to transform typescript node to  eslint node (which  is  estree, but  also not all the  same, eslint  don't have `StringLiteral`, they  use  `Literal` )

Reply:
> Using `MemberExpression` gave me type errors like this:
> 
> ```
> src/language-js/utils.js:912:6 - error TS2367: This condition will always return 'false' since the types '"MemberExpression"' and '"Literal"' have no overlap.
> 
> 912     (node.type === "Literal" && node.regex && node.regex.pattern) ||
> ```
> 
> I did change it to mix in `MemberExpression` and was able to get rid of the extra Expression properties.
> 
> I can give `@babel/types` a try. The eslint-typescript AST type seems to be specific to Typescript, or am I missing something there?

This node is  a  eslint version  `RegExpLiteral` , it's type is  `Literal` and  should come from `eslint-typescript`

Reply:
All our  js AST Nodes, none of them are  exactly estree, I'm not sure, but I think we need both  babel/eslint-typescript/flow

Reply:
I've been having trouble importing the types from eslint-typescript. Any pointers?

Reply:
Sorry, I don't have  this knowledge, can't help.

Reply:
https://github.com/typescript-eslint/typescript-eslint/blob/470174ad51fdb12d82129a896559075513f6c912/packages/types/src/ts-estree.ts#L141

This can't import?

Reply:
This works for me so far:

```diff
- * @typedef {import("estree").BaseNode} BaseNode
+ * @typedef {import("@typescript-eslint/types").TSESTree.Node} Node
+ * @typedef {import("@typescript-eslint/types").TSESTree.BaseNode} BaseNode
```

I will try it for the other JSDoc types, hang on.

Reply:
I don't think `BaseNode` is useful, just this `Node`, now import `Node` from `babel` and other parsers, union them all ad our `Node`.

I think it should work.